### PR TITLE
13297 - Deselect - way to cause deselection and stack-removal for a specific piece.

### PR DESCRIPTION
**Pull request by Cattlesquat**
image

There have been requests for the ability to deselect pieces w/ a key command. This implements that in a nice little trait. Optionally the deselected piece can also be popped out of its stack (useful when what you're doing is "dropping a piece off" from a moving stack). 

Provides all the normal trait features, and I have provided localization hooks.

If approved I'll provide the documentation.

### CODE REVIEW

**Review by uckelman — CHANGES_REQUESTED** [review]

File: vassal-app/src/main/java/VASSAL/counters/Deselect.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  public static final String ID = "deselect;";
+  protected KeyCommand[] command;
+  protected String commandName;
+  protected NamedKeyStroke key;
+  protected KeyCommand deselectCommand;
+  protected String description;
+  protected Boolean unstack;
+
+  public Deselect() {
+    this(ID + Resources.getString("Deselect.Deselect") + ";K;;false", null);
```
[the remark this answers is not in the document]
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/counters/Deselect.java:63]: That is actually the "default name" (localizable) of a default Deselect. The actual type is in ID I believe.

Mind you this thing was written before my most recent journeyman-course in Decent Sequence Encoder Separators.

[the remark this answers is not in the document]
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/counters/Deselect.java:63]: If we come up with some non-awful way to deal with "DELIMITER" in the Flare PR then I'd come fix this puppy right up.

[the remark this answers is not in the document]
Comment (uckelman) [reply vassal-app/src/main/java/VASSAL/counters/Deselect.java:63]: So, some other traits do this:
```
this(ID + "Delete;D", null);
```
`Delete` is the default command name there. That's _not_ translated. Is there an example of another trait which is translated?

@BrentEaston Thoughts?

File: vassal-app/src/main/java/VASSAL/counters/Deselect.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  
+  
+  /**
+   * Return Property names exposed by this trait
+   */
+  public List<String> getPropertyNames() {
+    ArrayList<String> l = new ArrayList<String>();
+    l.add(Properties.SELECTED);
+    return l;
+  }
```
[the remark this answers is not in the document]
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/counters/Deselect.java:214]: "Cannot resolve Constructor ArrayList(java.lang.String)"

**Comment by Cattlesquat** [comment]
Realized they're not supposed to get translated "in the editor". So the hardcode form is correct.

File: vassal-app/src/main/java/VASSAL/counters/Deselect.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/counters/Deselect.java:63]: You are correct - fixed & pushed.

**Comment by BrentEaston** [comment]
Yes, there is a method to translate menu commands when translating the module to another language.

However, I think your original point is valid, that if we are creating a new module using German as the base language,not English, then it is reasonable that the default Command names be supplied in German,

They can then be translated to English in the module if someone ones to make an English version.

This something I missed in my i18n guidelines. 

I am pretty sure there are other similar cases.

File: vassal-app/src/main/java/VASSAL/counters/Deselect.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (uckelman) [reply vassal-app/src/main/java/VASSAL/counters/Deselect.java:214]: There should be an `ArrayList<T>(T... args)` ctor. It's dumb that there's not. Never mind, the way you were doing it fine.

**Review by uckelman — APPROVED** [review]

**Comment by uckelman** [comment]
That's probably best to tackle for all traits at once.

**Pull request merged**